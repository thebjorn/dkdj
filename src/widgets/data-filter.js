

/*
 *   This is the top-level filter widget, which groups all sub-filters
 *   for a given data table/grid.
 */
// const _ = require('lodash');
// const dk = require('../boot');
// const Widget = require('../widgetcore/dk-widget.js');
// const forms = require('../forms');

import dk from "../dk-obj";
import {Widget} from "../widgetcore/dk-widget";
import {Template} from "../browser/dk-dom-template";
import {CheckboxSelectWidget, RadioSelectWidget, TriboolWidget} from "../forms/widgets";
import {zip_object} from "../collections";


export function list2data (lst) {
    if (Array.isArray(lst)) {
        const res = {};
        lst.forEach(function (item) {
            res[item] = item;
        });
        res.__order = lst;      // XXX: Not needed anymore
        return res;
    } else {
        return lst;
    }
}


/*
 *  The top level filter container/widget.
 *
 *  dk.filter.DataFilter.create_on(filterbox, {
 *      ...
 *      filters: {
 *          testleader: {
 *              label: 'Testleder',
 *              select_multiple: true,
 *              url: 'progress/all!get-testleader-values'
 *          },
 *          favyear: {
 *              label: 'Favoritt√•r',
 *              values: [2013, 2014, 2015, 2016],
 *              select_multiple: true
 *          },
 *          smile: {
 *              label: 'Smil',
 *              value: undefined,
 *              construct(location) {
 *                  const self = this;
 *                  const btn = $('<button/>').text(":-)");
 *                  this.widget = location.append(btn);
 *                  $(this.widget).on('click', function () {
 *                      self.value = "*SMILE*";
 *                      dk.publish(self.widget, 'change', self.widget);
 *                  });
 *              }
 *          }
 *      }
 *  });
 */
export class DataFilter extends Widget {
    constructor(...args) {
        super({
            // type: 'DataFilter',
            data: undefined,
            heading: 'Filter',
            filters: [],         // XXX shouldn't this be {} or null?
            dataset: null,
            structure: {
                header: {
                    title: {
                        text: 'Filter'
                    }
                },
                content: {}
            },

            template: {
                root: 'section',
                header: '<header/>',
                title: '<h2/>',
                content: '<div/>',

                // Each filter is contained in a .filterbox that is structured
                // similarly to the main filter section.
                filterbox: '<div/>',
                filterheader: '<header/>',
                filtertitle: '<h4/>',
                filtercontent: '<div class="filter content"/>'
            },
            
        }, ...args);
    }

    /*
     *  construct_filterbx is called for each filter definition in this.filters
     *  and creates the radio/check boxes for the filters, calls the data-set
     *  to fetch all filter values, etc.
     */
    construct_filterbx(location, filter, filter_name) {
        const self = this;
        const structure = {
            filterbox: {
                classes: [filter_name],
                filterheader: {
                    filtertitle: {
                        text: filter.label || filter_name
                    }
                },
                filtercontent: {}
            }
        };
        const templ = Template(structure, this.template);
        const dom = templ.append_to(location, this, location);

        if (!filter.construct) {
            // Filters that don't construct themselves are constructed either as single or
            // multiple selects. Default is single-selection (i.e. radio selects).
            let widgetclass = filter.select_multiple ? CheckboxSelectWidget : RadioSelectWidget;
            if (filter.tribool) {
                widgetclass = TriboolWidget;
            }

            filter.construct = function (loc) {
                // this here refers to the filter, which we are attaching a new method to..
                const me = this;

                // noinspection JSPotentiallyInvalidUsageOfClassThis
                this.widget = widgetclass.create_inside(loc, {
                    data: list2data(this.values),
                    name: filter_name,
                    label: me.label
                });

                if (!this.widget.data) {
                    self.dataset.get_filter_data(filter_name, this.widget.FN('draw'));
                }
            };
        }
        filter.construct(location.filterbox.filtercontent);
        if (filter.widget) {
            // filters with their own `construct()` don't necessarily have `.widget`
            dk.on(filter.widget, 'change').run(function () {
                dk.debug('filter-change', self.values());
                if (self.data) self.data.set_filter(self.values());
                dk.publish(self, 'filter-change', self.values());
            });
        }
        return dom;
    }

    construct() {
        this.filters.forEach((value, key) => {
            //console.log('key:', key, 'value:', value);
            this.construct_filterbx(this.content, value, key);
        });
    }

    values() {
        const self = this;
        const keys = Object.keys(this.filters);
        const vals = keys.map(function (filter_name) {
            const filter = self.filters[filter_name];
            if (filter.value) {
                // filter has defined a value on itself..
                return filter.value;
            }
            if (filter.widget && filter.widget.jquery) {
                // found a jQuery object, get its .val()
                return filter.widget.val();
            }
            if (filter.widget && filter.widget.value) {
                // found a dk.form object
                return filter.widget.value;
            }
        });
        // dk.dir(_.object(keys, vals));
        return zip_object(keys, vals);
    }
}
