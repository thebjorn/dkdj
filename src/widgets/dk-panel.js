
import {Widget} from "../widgetcore/dk-widget";
import css from "../browser/dk-css";
import dk from "../dk-obj";
import browser from "../browser/browser-version";
import {dkicon, fa4_icon, icon} from "./dk-icon-library";
import {dkconsole} from "../lifecycle/dkboot/dk-console";


export class PanelWidget extends Widget {
    constructor(props) {
        super({
            // type: 'PanelWidget',
            width: undefined,
            height: undefined,
            collapsible: true,
            direction: 'up',   // or 'left'
            blind_duration: 0.1,
            rotate_duration: 0.1,
            status: 'open',
            icons: {
                collapse: 'folder-open-o:fw',
                expand: 'folder:fw',
                color: '#888'
            },

            overflow: 'resize',   // 'hidden'|'scroll'|'resize'
            kind: 'panel-default',

            structure: {
                classes: ['dk-panel', 'panel'],
                css: {
                    position: 'relative',
                    overflow: 'hidden'
                },

                header: {
                    classes: ['panel-heading'],
                    title: {
                        classes: ['panel-title'],

                        collapseicon: {
                            css: {cursor: 'pointer'}
                        },
                        headingtext: {
                            template: 'span'
                        }
                    }
                },
                panel_body: {
                    css: {overflow: 'auto'}
                },
                footer: {
                    classes: ['panel-footer'],
                    create: false
                }
            }
        }, props);

        if (this.panel && this.panel.height && typeof this.panel.height === 'string') {
            this.height = parseInt(this.panel.height, 10);
        }
    }

    construct_headingtext(htxt) {
        if (this.title) {
            htxt.text(this.title);
        }

        if (!this.icon) this.icon = this.widget('.collapseicon > dk-icon')[0];
        if (!this.icon) {
            const icon = dk.$('<dk-icon/>').attr('value', this.icons.collapse);
            this.icon = icon[0];
            this.widget('.collapseicon').append(icon);
        }
        return htxt;
    }

    construct_panel_body(body) {
        return body;
    }

    construct() {
        if (this.widget().hasClass('collapse-up'))      this.status = 'collapse-up';
        if (this.widget().hasClass('collapse-left'))    this.status = 'collapse-left';
        if (this.height) {
            this.panel_body.outerHeight(this.height - this.header.outerHeight() - this.footer.outerHeight());
        }
        if (!this.icon)  this.icon = this.widget('.collapseicon > dk-icon')[0];
    }

    set_icon(ikn) {
        this.icon.value = ikn;
    }

    _size(item) {
        return {
            height: item.outerHeight(),
            width: item.outerWidth()
        };
    }

    collapse_up() {
        const self = this;
        const wsize = this._size(this.widget());
        const hsize = this._size(this.header);

        this.widget().addClass('collapse-up');
        const orig_maxheight = css.maxheight(this.widget(), {
            height: hsize.height,
            duration: self.blind_duration,
            done() {
                self.__collapsing = false;
            }
        });

        this._expand = function () {
            self.widget().removeClass('collapse-up');
            css.maxheight(self.widget(), {
                height: wsize.height,
                duration: self.blind_duration,
                done() {
                    self.widget().css('max-height', orig_maxheight);
                    self.__collapsing = false;
                }
            });
        };
    }

    collapse_left() {
        const self = this;
        const wsize = this._size(this.widget());
        const hsize = this._size(this.header);
        let rotation;

        const orig_maxheight = css.maxheight(this.widget(), {
            height: hsize.height,
            duration: self.blind_duration,
            done: function () {
                rotation = css.tilt(self.widget(), {
                    direction: 'bottom left',
                    duration: self.rotate_duration,
                    done: function () {
                        self.__collapsing = false;
                        // noinspection JSSuspiciousNameCombination
                        self.trigger('collapse-done', {
                            height: hsize.width,
                            width: hsize.height
                        });
                    }
                });
                self.widget().addClass('collapse-left');
            }
        });

        this._expand = function () {
            css.unrotate(dk.update(rotation, {
                done: function () {
                    css.maxheight(self.widget(), {
                        height: wsize.height,
                        duration: self.blind_duration,
                        done: function () {
                            self.widget().css('max-height', orig_maxheight);
                            self.__collapsing = false;
                            self.trigger('expand-done', self);
                        }
                    });
                }
            }));
        };
    }

    collapse(direction) {
        this.direction = direction || this.direction;
        this['collapse_' + this.direction]();
        this.icon.value = this.icons.expand;
        this.trigger('resized', this);
        this.trigger('collapse', this);
    }

    expand() {
        this.widget().removeClass('collapse-' + this.direction);
        this.icon.value = this.icons.collapse;
        if (this._expand) {
            this._expand();
            this._expand = null;
        }
        this.trigger('resized', this);
        this.trigger('expand', this);
    }

    handlers() {
        const self = this;
        if (this.collapsible) {
            //this.__collapsing = new Date();

            this.header.on('click dblclick', function () {
                if (self.__collapsing) return;
                self.__collapsing = true;
                const action = self.collapsed ? 'expand' : 'collapse';
                self.trigger(action + '-start', self);
                self[action]();
                self.collapsed = !self.collapsed;
            });
        }
    }
}


/**
 * Usage::
 * 
 *     <dk-panel>
 *         <h3>panel title</h3>
 *         PANEL BODY
 *     </dk-panel>
 */
if (typeof customElements !== 'undefined') customElements.whenDefined('dk-panel').then(() => {
    // console.log("DK_PANEL_DEFINED:");
});

let _dk_panel_template = null;
function get_dkpanel_template() {
    if (!_dk_panel_template) {
        _dk_panel_template = document.createElement('template');
        _dk_panel_template.innerHTML = `
            <link rel="stylesheet" href="https://netdna.bootstrapcdn.com/bootstrap/3.1.0/css/bootstrap-theme.min.css">
            <link rel="stylesheet" href="https://static.datakortet.no/font/fa470/css/font-awesome.css">
            <link rel="stylesheet" href="/dkjs/dkjs/static/dkjs/js/dkcss.css" type="text/css">
<!--            <link rel="stylesheet" href="//static.datakortet.no/dkjs/dkcss.fa530d8f7e49451dd630.css">-->
            <style>
                :host {
                    display: block;
                }
            </style>
            <div class="PanelWidget dk-panel panel panel-default">
                <header class="panel-heading">
                    <div class="panel-title">
                        <span id="icon" class="collapseicon" style="cursor:pointer;"></span>
                        <span id="title-text" class="headingtext"></span>
                    </div>
                </header>
                <slot id="panel-body" name="panel-body"></slot>
            </div>
            <slot id="panel-content"></slot>
        `;
    }
    return _dk_panel_template.content.cloneNode(true);
}
let _panel_counter = 1;

if (typeof customElements !== 'undefined') {
    if (browser.name === "msie" && browser.versionNumber <= 11) {
        customElements.define('dk-panel', class extends HTMLElement {
            constructor() {
                super();
                if (!this.id) this.id = `dk-panel-ms-${_panel_counter++}`;
            }
            connectedCallback() {
                dk.$(this).css('display', 'block');
                const $header = dk.$(this).find('>:header:first-child');
                const $img = $header.find('>img:first-child');
                $header.detach();
                dk.$(this).wrapInner('<div class="panel-body"/>');
                
                this._widget = PanelWidget.create_on('#' + this.id, {
                    title: $header.text()
                });
                try {
                    this._widget.footer.hide();
                } catch (e) {}
                if ($img.length && this._widget.header) {
                    this._widget.header.title.collapseicon.empty().append($img);
                }
            }
        });
    } else {
        
        // normal browsers..
        customElements.define('dk-panel', class extends HTMLElement {
            constructor() {
                super();
                this._collapsed = false;
                this._icon_open = 'folder-open-o:fw';
                this._icon_closed = 'folder:fw';
                this._dkicon = null;
                
                const shadowRoot = this.attachShadow({mode: 'open'});
                shadowRoot.append(get_dkpanel_template());
                this._icon = shadowRoot.getElementById('icon');
                this._panel_title_text = shadowRoot.getElementById('title-text');
                this._slot = shadowRoot.getElementById('panel-content');
                this._panel_body_slot = shadowRoot.getElementById('panel-body');
                this._slot_changed = false;

                this._slot.addEventListener('slotchange', () => this._on_slot_change());
                this._icon.addEventListener('click', (e) => this.toggle(e));
            }

            _on_slot_change() {
                const add_body_panel = () => {
                    const panel_body = document.createElement('div');
                    panel_body.classList.add('panel-body');
                    // this._panel.append(panel_body);
                    this.append(panel_body);
                    panel_body.setAttribute('slot', 'panel-body');
                    return panel_body;
                };

                if (!this._slot_changed) {  // this function is called twice..
                    this._slot_changed = true;

                    const header = this.querySelector('h1,h2,h3,h4,h5,h6');  // :header
                    this._panel_title_text.textContent = header.textContent;
                    
                    const img = header.querySelector('img:first-child');
                    if (img) {
                        // img.setAttribute('slot', 'icon');
                        this._icon.append(img);
                    } else {
                        this._dkicon = icon(this.hasAttribute('collapsed') ? this._icon_closed : this._icon_open);
                        this._icon.append(this._dkicon);
                        // this._dkicon = dkicon.create_inside(this._icon, {value: this._icon_open});
                    }
                    this.removeChild(header);

                    let current_panel_body = null;
                    if (this.children.length === 0 && this.childNodes.length > 0) {  // only text nodes
                        current_panel_body = add_body_panel();
                        while (this.childNodes.length > 0 && this.childNodes[0] !== current_panel_body) {
                            current_panel_body.appendChild(this.childNodes[0]);
                        }
                    }                    
                    [...this._slot.assignedElements()].forEach(n => {
                        switch (n.tagName) {
                            case 'TABLE':
                            case 'UL':
                                n.setAttribute('slot', 'panel-body');
                                current_panel_body = null;
                                break;
                            default:
                                if (!current_panel_body) current_panel_body = add_body_panel();
                                current_panel_body.append(n);                                
                        }
                    });
                }
            }
            
            connectedCallback() {
                this.collapsed = this.hasAttribute('collapsed');
            }
            
            static get observedAttributes() {
                return [
                    'collapsed',
                ];
            }
            
            attributeChangedCallback(attrname, oldval, newval) {
                newval = newval === "";  // true
                oldval = oldval === "";
                // console.log(`attrchange: ${attrname} #${this.id} "${oldval}"->"${newval}"`);
                if (attrname === 'collapsed' && oldval != newval) {
                    // console.log(`performing-attrchange: ${attrname} #${this.id} "${oldval}"->"${newval}"`);
                    this.collapsed = newval;
                }
            }

            get collapsed() { return this._collapsed; }
            set collapsed(v) {
                if (!!v !== this._collapsed) {
                    if (v) {
                        this.collapse();
                    } else {
                        this.expand();
                    }
                }
            }
            
            collapse() {
                if (!this.collapsed) {
                    this._collapsed = true;
                    dk.$(this._panel_body_slot).hide();
                    this.setAttribute('collapsed', "");
                    if (this._dkicon) {
                        fa4_icon(this._dkicon, this._icon_closed);
                        // this._dkicon = icon(this._icon_closed);
                    }
                }
            }
            expand() {
                if (this.collapsed) {
                    this._collapsed = false;
                    dk.$(this._panel_body_slot).show();
                    this.removeAttribute('collapsed');
                    if (this._dkicon) {
                        fa4_icon(this._dkicon, this._icon_open);
                        // this._dkicon = icon(this._icon_open);
                    }
                }
            }
            toggle() {
                if (this.collapsed) {
                    this.expand();
                } else {
                    this.collapse();
                }
            }
        });

    }
}
