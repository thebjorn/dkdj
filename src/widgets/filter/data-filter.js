/*
 *   This is the top-level filter widget, which groups all sub-filters
 *   for a given data table/grid.
 */

import dk from "../../dk-obj";
import {zip_object} from "../../collections";
import {UIWidget} from "../../widgetcore/ui-widget";
import is from "../../is";
import {dkconsole} from "../../lifecycle/dkboot/dk-console";
import {CustomFilterDef, SelectMultipleFilterDef, SelectOneFilterDef} from "./filter-defs";



/*
 *  The top level filter container/widget.
 *
 *  dk.filter.DataFilter.create_on(filterbox, {
 *      ...
 *      filters: {
 *          testleader: {
 *              label: 'Testleder',
 *              select_multiple: true,
 *              url: 'progress/all!get-testleader-values'
 *          },
 *          favyear: {
 *              label: 'Favoritt√•r',
 *              values: [2013, 2014, 2015, 2016],
 *              select_multiple: true
 *          },
 *          smile: {
 *              label: 'Smil',
 *              value: undefined,
 *              construct(location) {
 *                  const self = this;
 *                  const btn = $('<button/>').text(":-)");
 *                  this.widget = location.append(btn);
 *                  $(this.widget).on('click', function () {
 *                      self.value = "*SMILE*";
 *                      dk.publish(self.widget, 'change', self.widget);
 *                  });
 *              }
 *          }
 *      }
 *  });
 */
export class DataFilter extends UIWidget {
    constructor(...args) {
        super({
            // type: 'DataFilter',
            data: undefined,
            heading: 'Filter',
            filters: {},
            dataset: null,
            structure: {
                header: {
                    title: {
                        text: 'Filter'
                    }
                },
                content: {}
            },

            template: {
                root: 'section',
                header: '<header/>',
                title: '<h2/>',
                content: '<div/>',

                // Each filter is contained in a .filterbox that is structured
                // similarly to the main filter section.
                filterbox: '<div/>',
                filterheader: '<header/>',
                filtertitle: '<h3/>',
                filtercontent: '<div class="filter content"/>'
            },
            
        }, ...args);
        
        this._filter_data = this._get_filter_data(this.filters);
    }

    /**
     *  Convert the user supplied `.filters` to a list of FilterDefBase 
     *  subclasses, which can (later) be combined with a location to create
     *  a Widget.
     *  
     *  filterdata filterdef is either
     *    
     *    (a) { select_multiple: true, values: .., ... }
     *    (b) { construct(location) {}, ... }
     *    (c) { values: .., ... }
     *   
     *  For option (b), the object must 
     *  
     *    (i) create a `this.widget` property (the filter system listens
     *        for notifications on this widget.
     *   (ii) set `this.value` to the correct filter value (this can be a
     *        compound value, but must be _JSON_ serializable).
     *  (iii) call `dk.trigger(this.widget, 'change', this.widget)` whenever
     *        the filter value changes.
     *        
     */
    _get_filter_data(filters) {
        return Object.entries(this.filters).map(([filtername, filterdef]) => {
            filterdef.name = filtername;
            if (is.isProps(filterdef)) {
                if (filterdef.select_multiple) {  // (a)
                    // console.info("SELECT_MULTIPLE", filterdef);
                    return [filterdef, SelectMultipleFilterDef];
                } else if (filterdef.construct_filter) {  // (b)
                    // console.info("CUSTOMFILTER", filterdef);
                    return [filterdef, CustomFilterDef];
                } else {
                    // console.info("SELECT_ONE", filterdef);
                    return [filterdef, SelectOneFilterDef];
                }
            } else {
                dkconsole.warn("filter def is not a properties object..?");
                // return [filterdef, null]; 
            }
        });
    }

    /**
     * Create widgets, and append to `this.content`, for all pairs in 
     * `this._filter_data`. Set it up so any `change` trigger on the filters 
     * triggers a `filter-change` event on the filter (with `this.values()` 
     * as the only parameter.
     */
    construct() {
        this.filters = {};
        this._filter_data.forEach(([filterprops, filtercls]) => {
            filterprops.datafilter = this;
            const res = filtercls.append_to(this.content, filterprops);
            dk.on(res, 'change', (...args) => {
                if (this.data) this.data.set_filter(this.values());
                this.trigger('filter-change', this.values());
            });
            this.filters[filterprops.name] = res;
        });
    }

    values() {
        const keys = Object.keys(this.filters);
        const vals = keys.map(filter_name => {
            const filter = this.filters[filter_name];
            if (filter.value) {
                // filter has defined a value on itself.. (e.g. `CustomFilterDef`'s
                return filter.value;
            }
            if (filter.input && filter.input.jquery) {
                // found a jQuery object, get its .val()
                return filter.input.val();
            }
            if (filter.input && filter.input.value) {
                // found a dk.form object
                return filter.input.value;
            }
        });
        return zip_object(keys, vals);
    }
}
