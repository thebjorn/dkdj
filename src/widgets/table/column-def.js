
import Class from "../../lifecycle/coldboot/dk-class";
import dk from "../../dk-obj";
import {value as format_value} from "../../data/datacore/dk-format"
import widgetmap from "../../widgetcore/dk-widgetmap";
import {TextInputWidget} from "../../forms/widgets";


export class ColumnDef extends Class {
    constructor(...args) {
        super({
            isa:         'dk.table.ColumnDef',
            name:        '',                        // name of column (eg. for computed cols)
            field:       '',                        // field definition {name:.., pos:.., string:..,...}
            label:       '',                        // text to put in th element
            sortable:    undefined,                 // sortable override
            description: '',                        // description
            align:       undefined,                 // cell alignment
            format:      format_value,  // formatting function
            type:        undefined,                 // data type of column values
            widget_type: TextInputWidget,     // widget to use when editing a cell value
            empty:       '',                        // value to use for empty cells
            table:       undefined,                 // the DataTable (subclass) we belong to
            _alignment_map: {
                int: 'right'
            }
        }, ...args);
        this.name = this.name || this.field;
    }

    alignment() {
        return this.align || this._alignment_map[this.type] || 'left';
    }

    get_value(record) {
        //dk.info("getting value from columndef", record, this, record[this.name]);
        if(record)
            return record[this.name];
    }

    /*
     *  Update self with info from the datasource.
     */
    bind_to_field(field) {
        this.field = field;
        if (!this.label) this.label = field.label || field.name;
        if (!this.type) this.type = field.type;

        let wtype;
        if (field.widget) {
            // wtype = widgetmap.get(field.widget);
            wtype = widgetmap[field.widget];
        } else {
            wtype = widgetmap[this.type];
        }

        if (wtype === undefined) {
            // dk.warn("Unknown widget: ", field.widget, field);
            wtype = TextInputWidget;
        }
        this.widget_type = wtype;
        this.widget_data = field.data;
        this.widget_url = field.url;
    }
}
