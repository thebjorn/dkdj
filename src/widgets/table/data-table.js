/*
 *  A DataTable is the normal table representation of a DataSource,
 *  i.e. records in rows, with each field in its own column (with a column
 *  header).
 */

import dk from "../../dk-obj";
import {TableLayout} from "../../layout/table-layout";
import {Widget} from "../../widgetcore/dk-widget";
import {TableRow} from "./table-row";
import is from "../../is";
import {ColumnDef} from "./column-def";
import {TableHeader} from "./table-header";
import {DataSet} from "../../data/dk-dataset";
import {AjaxDataSource} from "../../data/source/dk-ajax-datasource";
import {dkwarning} from "../../lifecycle/coldboot/dkwarning";


export class DataTableLayout extends TableLayout {
    constructor(widget, location, template, structure) {
        super(widget, location, template, structure);
        this.thead_tr = this.add_header_row();
    }
    
    add_thead_th(attrs) {
        const th = this.make('thead_th', attrs);
        this.thead_tr.appendln(th);
        return th;
    }
}


export class DataTable extends Widget {
    constructor(...args) {
        super({
            // type: 'datatable',
            dklayout: DataTableLayout,
            TableRow: TableRow,
            current_pagenum: 0,
            translate: {},  // l10n

            structure: {
                thead: {},
                tfoot: {},
                tbody: {}
            },

            template: {
                root: 'table',
                table: '<table/>',
                thead: '<thead/>',
                thead_tr: '<tr/>',
                thead_th: '<th/>',
                tbody: '<tbody/>',
                tbody_tr: '<tr/>'
            },

            table_data: null,
            state: undefined,

            // The columns attribute should be a property object that defines the visible columns:
            //
            //      columns: {
            //          colnam1: {}
            //          colname2: {..}
            //      ..
            //
            // this.columns is converted to an array of dk.table.ColumnDef, and this.column to a
            // dict from column name to dk.table.ColumnDef.
            columns: null,
            column: {},

            filters: {},

            // should table be sortable?
            sortable: true,

            // selector of dom element to bless as a download link (e.g.
            // <a id="foo" data-download="filename.csv">download</a>
            // =>  download: '#foo',
            download: undefined,

            header: null,
            rows: null,

            column_order: null,
            column_defs: null,            
        }, ...args);

        const self = this;
        this.rows = [];
        this.header = [];   // th cell objects

        this.column = {};
        this.column_order = [];
        if (!!this.datasource && this.table_data === null) {
            const dsprops = {datasource: this.datasource};
            if (this.pagesize != null) dsprops.pagesize = this.pagesize;
            if (this.orphans != null) dsprops.orphans = this.orphans;
            if (this.pagenum != null) dsprops.pagenum = this.pagenum;
            this.table_data = new DataSet(dsprops);
        }
        if (this.table_data === null) {
            this.data_url = (this.dk && this.dk.url)? this.dk.url : "";
            //dk.info("URL", this.data_url);
            // if data is not defined, then create one
            this.table_data = new DataSet({
                // default data source is AjaxDataSource
                datasource: new AjaxDataSource({
                    // default url is current url
                    url: self.data_url
                })
            });
        }
        if (!this.data_url) this.data_url = this.table_data.datasource.url;

        //        this.state = dk.State.create({
        //            state: this.table_data.current_pagedef,
        //            engine: dk.HashStorage.create({})
        //        });
        //this.state.restore();
    }

    construct() {
        // this.state = dk.page.hash.substate(this.id);

        if (this.download) {
            const $download = dk.$(this.download);
            const filename = $download.attr('data-download') || 'filename.csv';
            dk.$(this.download).attr('href', this.data_url + "!get-records?fmt=csv&filename=" + filename);
        }
    }

    /*
     *  Connect to a pager widget.
     */
    set_pager(pwidget) {
        // const self = this;
        if (this.table_data.page) {
            pwidget.set_pagecount(this.table_data.page.recordset.meta.pagecount);
            pwidget.select_page(this.table_data.page.query.pagenum);
        }

        dk.on(pwidget, 'select-page', (...args) => this.table_data.get_pagenum(...args));
        dk.on(this.table_data, 'fetch-info', (info, query) => {
            pwidget.set_pagecount(info.pagecount);
            pwidget.select_page(query.pagenum);
        });
    }

    filter_values() {
        return {};
    }

    handlers() {
        //$(this.table_data).on('fetch-data', _.bind(this.draw, this));
        //$bind('fetch-data@data -> draw@me', {data: this.table_data, me: this});
        dk.on(this.table_data, 'fetch-data-start', () => this.start_busy());
        dk.on(this.table_data, 'fetch-data', (...args) => this.draw(...args));
        dk.on(this.table_data, 'fetch-data', (...args) => this.end_busy(...args));
    }

    set_sort(sortcol) {
        if (!sortcol.sortable) return;
        if (this.sortable) {
            this.header.forEach(function (col) {
                if (col.sortable && col !== sortcol) col.sort_icon.clear_sort();
            });
            sortcol.sort_icon.toggle_sort_direction();
            this.table_data.set_sort(sortcol.name, sortcol.sort_icon.direction);
            dk.trigger(this, 'sort-order', sortcol.name, sortcol.sort_icon.direction, this);
        }
    }

    set_search(terms) {
        this.table_data.set_search(terms);
        if (this.state) this.state.set('pagedef', this.table_data.current_pagedef);
    }

    set_filter(vals) {
        this.table_data.set_filter(vals);
    }

    get_xy(colnum, rownum) {
        return this.rows[rownum].cells[colnum];
    }

    set_xy(colnum, rownum, val, html, title) {
        const cell = this.get_xy(colnum, rownum);
        cell.set_value(val, html, title);
        return cell;
    }

    values() {
        const res = [];
        this.rows.forEach(function (row) {
            const r = [];
            row.cells.forEach(function (cell) {
                r.push(cell.widget().text());
            });
            res.push(r);
        });
        return res;
    }

    /*
     *  Convert this.columns to an array of basic property objects.
     *
     *    {foo: {label: 'Foo', ..}} ==> [{name: 'foo', label: 'Foo', ...}, ...]
     */
    _get_columndefs(fieldlst) {
        const self = this;
        let columndefs;
        const fields = {};  // field.name -> field
        fieldlst.forEach(function (f) { fields[f.name] = f; });

        // make sure this.columns is an array of pojos
        if (is.isFunction(this.columns)) columndefs = this.columns(this.table_data.fields);
        else if (this.columns === null) columndefs = fieldlst;
        else if (!Array.isArray(this.columns)) {
            // the columns property has been defined, convert it to array
            columndefs = Object.keys(this.columns).map(name => {
                const col = this.columns[name];
                col.name = name;
                return col;
            });
        } else {
            dkwarning(`Expected {field1: {...}, field2: {...}, etc.}. columns property of unknown type: ${this.columns}`);
        }

        // convert the pojos to ColumnDef objects
        return columndefs.map(function (coldef, colnum) {
            const field = fields[coldef.name];
            const tmp = {};
            if (field) {
                Object.assign(tmp, {
                    fieldpos: field.pos,
                    fieldtype: field.type,
                    sortable: field.sortable,
                    label: field.label,
                    help_text: field.help_text,
                    field_widget: field.widget,
                    field_data: field.data
                });
            }
            Object.assign(tmp, coldef);
            Object.assign(tmp, {
                colnum: colnum,
                table: self
            });
            if (tmp.sortable === undefined) {
                tmp.sortable = self.sortable;
            }
            const cdef = new ColumnDef(tmp);
            if (fields[cdef.name]) cdef.bind_to_field(fields[cdef.name]);
            self.column[cdef.name] = cdef;
            self.column_order.push(cdef.name);
            return cdef;
        });
    }

    /*
     *  Draw the column headers.
     */
    draw_header(dataset) {  // must be a separate method so VDataTable can inherit..
        if (!this.drawn_header) {
            const self = this;
            // NOTE: we have data here!
            this.columns = this._get_columndefs(dataset.page.fields);

            this.columns.forEach(function (coldef) {
                const th = TableHeader.create(self.layout.add_thead_th(), coldef);
                self.header.push(th);
            });
            self.drawn_header = true;  // only draw header once.
        }
    }

    draw(dataset) {
        const self = this;

        if (!dataset) {   // NOTE: data source consumer pattern..?  (on: data -> draw?)
            this.table_data.get_page({
                pagenum: this.current_pagenum
            });
        } else {
            self.draw_header(dataset);
            dk.trigger(self, 'draw-start', self);
            self.layout.clear_body();
            self.rows = [];

            dataset.page.records.forEach(function (record, rownum) {
                const tr = self.TableRow.create(self.layout.add_row_to('tbody'), {
                    rownum: rownum,
                    record: record,
                    table: self
                });
                self.rows.push(tr);
            });
            dk.trigger(self, 'draw-end', self);
        }
    }
}
