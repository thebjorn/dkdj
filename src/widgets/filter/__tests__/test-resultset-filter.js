import $ from "jquery";
import page from "../../../widgetcore/dk-page";
import {PanelWidget} from "../../dk-panel";
import {ResultSet} from "../../table/resultset/resultset";
import {DataTable} from "../../table/data-table";
import {DataSet} from "../../../data/dk-dataset";
import {DataFilter} from "../data-filter";
import utidy from "../../../browser/dk-html";
import {ArraySource} from "../../../data/source/dk-array-datasource";


test("resultset-filter-options", () => {
    document.body.innerHTML = `
        <div id="work"></div>
    `;
    const work = $('#work');
    page.initialize(document);
    
    let set_filter_vals = null;
    let gfd = null;

    class MyArraySource extends ArraySource {
        get_filter_data(filter_name, returns) {
            gfd = 'called';
            returns({
                options: [1, 2, 3]
            });
        }
    }
    
    const ds = new MyArraySource([
        {f1: 'hello', f2: 42}
    ]);

    const dset = new DataSet({
        datasource: ds
    });
    

    class ProgressResultFilterPanel extends PanelWidget {
        constructor(...args) {
            console.log("ProgressResultFilterPanel", args);
            super({
                direction: 'left',
                title: 'Filter',
                dataset: null,
            }, ...args);
            console.log('progresfiltpan:', this.datasource, this.dataset);
        }

        construct_panel_body(body) {
            const self = this;
            const filterbox = $('<div/>');
            body.append(filterbox);

            this.datafilter = DataFilter.create_on(filterbox, {
                structure: {
                    content: {}
                },
                dataset: self.dataset,
                filters: {
                    testcenter: {
                        label: 'Testsenter',
                        select_multiple: true,
                    }
                }
            });
            return body;
        }
    }


    ResultSet.create_on(work, {
        construct_filter: function (location, dataset) {
            // return DataFilter.create_on(location, {
            //     structure: {
            //         content: {}
            //     },
            //     datasource: ds,
            //     filters: {
            //         testcenter: {
            //             label: 'Testsenter',
            //             select_multiple: true,
            //             // values: [1,2,3]
            //             // url: './!get-testcenter-values'
            //         }
            //     }
            // });
            return ProgressResultFilterPanel.create_on(location, {
                // datasource: ds,
                dataset: dataset
            });
        },

        construct_table: function (location, downloadwidget) {
            console.info("constructing table...", location, downloadwidget);
            return DataTable.create_on(location, {
                classes: ['table table-bordered table-hover table-condensed'],
                download: downloadwidget,

                table_data: new DataSet({
                    pagesize: 20,
                    datasource: ds
                }),
                columns: {
                    pk: {}
                }
            });
        }
    });

    expect(utidy(work.html())).toEqual(utidy(`

    `));
});
