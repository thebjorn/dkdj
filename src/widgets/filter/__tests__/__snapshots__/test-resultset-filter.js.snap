// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resultset-filter-options 1`] = `[Function]`;
